tools/close_by_signature.py

```diff
@@ -0,0 +1,243 @@
+    )
+    dry_run_args.add_argument(
+        '-t', '--dry-run-temp',
+        action='store_true',
+        help=f'Dry run. Don\'t update tickets. Write output to a temp file'
+    )
+
+    return parser.parse_args()
+
+
+class IssueData:
+
+    def __init__(self, issue, signature, comment):
+        self.issue = issue
+        self.signature = signature
+        self.comment = comment
+        self.root_issue = None
+
+
+def get_signatures(signature_types, jira):
+    signatures = filter(
+        lambda x: x.__name__.lower() in signature_types,
+        SIGNATURES,
+    )
+    return [sign(jira) for sign in signatures]
+
+
+def issues_by_signature_generator(jira, signatures, issues):
+    for issue in issues:
+        if issue.fields.status.name in ('Done', 'Obsolete'):
+            continue
+
+        comments = jira.comments(issue.key)
+        for sign in signatures:
+            comment = sign.find_signature_comment(issue.key, comments)
+            if comment is None:
+                continue
+
+            yield IssueData(issue, sign, comment)
+            break
+
+
+def filter_issues_by_message_generator(jira, issues_data_generator, messages):
+    msgs_to_search_for = []
+    msgs_to_root_issue = {}
+    for msg in messages:
+        if ':' not in msg:
+            msgs_to_search_for.append(msg)
+            continue
+
+        original_msg = msg
+        root_issue_key, msg = msg.split(':', 1)
+        root_issue = jira.issue(root_issue_key)
+        if root_issue is None:
+            msgs_to_search_for.append(original_msg)
+            continue
+
+        msgs_to_root_issue[msg] = root_issue
+        msgs_to_search_for.append(msg)
+
+    for issue_data in issues_data_generator:
+        for msg in msgs_to_search_for:
+            if msg not in issue_data.comment.body:
+                continue
+
+            issue_data.root_issue = msgs_to_root_issue.get(msg)
+            yield issue_data
+            break
+
+        if not msgs_to_search_for:
+            yield issue_data
+
+
+def close_and_link_issues(jira, issues_data_generator):
+    for issue_data in issues_data_generator:
+        if issue_data.root_issue:
+            link_issue_to_root_issue(
+                jira=jira,
+                issue=issue_data.issue,
+                root_issue=issue_data.root_issue,
+            )
+
+        logger.debug('Closing issue with key=%s signature=%s comment=%s',
+                     issue_data.issue.key, type(issue_data.signature),
+                     issue_data.comment.body)
+
+        if dry_run_stdout is None:
+            # set issue status to 'Done'
+            jira.transition_issue(issue_data.issue, '31')
+        else:
+            dry_run_stdout.write(f'Closed issue key={issue_data.issue.key} '
+                                 f'signature={type(issue_data.signature)} '
+                                 f'comment={issue_data.comment.body}\n')
+            dry_run_stdout.flush()
```

Comment:
I think you can do ``print("something", file=...)`` instead. Also provide it with the ``flush`` argument if needed